Complete rewrite for HTML5 spec compatiblity

reppners opened the pull request:
Hey there,

it started out as an attempt to make this shim work with an existing html5 drag and drop library we use in our angular project and ended up as a rewrite that is as close as possible to the HTML5 spec and support for all mobile browsers, making it a polyfill that aims to fill the gap for all browsers currently not implementing the spec natively.

Are you interested in pulling this back in?

jmeas commented:
Dang.

:clap:

It's weird how those little projects can turn into something much larger :)

I haven't had the opportunity to look this over, but I am 100% interested in expanding the scope of this project to be more useful. @timruffles , what's your take?

timruffles commented on .gitignore:
```diff
@@ -0,0 +1,55 @@
+# Created by .ignore support plugin (hsz.mobi)
```
Place `.gitignore` lines that are about your setup in your `.gitignore_global`; place stuff that's essential to the project in its `.gitignore`.

timruffles commented:
Looks interesting, will have a look ASAP!

reppners commented:
Thanks for taking the time. I removed the template stuff from the *.gitignore

Yesterday I was in a hurry and pressed enter too soon, leaving the pull request without any message so I did just a quick edit afterwards, not explaining all the things I did and why but I think its best you leave a comment here on any things that look strange to you.

I did a few rounds of explorative testing, comparing desktop Safari, Firefox and Chrome output directly with a connected iPad(iOS8), Nexus(Lollipop) and also have a MS Surface at hand that I used to convince Chrome with enabled touches to play with me. Some bugs are still hiding but for now it looks quite solid to me. Only thing I couldn't test right now is MS Edge but a VM with Windows 10 is already setup and IEMobile.

Right now I just added a few bugfixes and a first stab for a feature: scrolling when dragging close to a viewport edge. I'm also thinking about combining it with an optional zoom out/zoom in like some mobile browsers do already when something is dragged close to the right edge of the viewport unless you stop them from doing so with the meta viewport tag by set initial, min and max zoom to 1.

reppners commented:
Continued to improve and fix bugs.

Most recent addition is a debug mode that highlights the dispatched events, current drop target and immediate user selection. Check out the lightshow: http://reppners.github.io/ios-html5-drag-drop-shim/spec-compliance/

Right now the demo page is not really a demo page but rather my development playground so forgive me any rough edges :D

timruffles commented:
Looks amazing work :) Amazing job on getting into the detail of the spec.

Let me know what you think needs to be done to make it ready for release, and I'll pitch in where I can. I guess the most important bit is to identify any breaking changes from 1.0?

reppners commented:
Thanks :) 

Today I fixed the last bugs that prevented me from using it in a bigger project and so far I did not discover any greater issues, but a bit more testing/beta phase would be a good thing.

Detecting breaking changes is very important. Has anybody suggestions on how to tackle this? Write down a list of tests? Automated tests?

Thinking about a release a few other things come to my mind:
*    Does the term "mobile" still apply? Because desktop Chrome is also patched and this polyfill aims to be universal.
*    Any concerns about the file size? 
     It has grown quite a bit, maybe we can try the best to reduce it, by excluding drop zone support (does any DnD supporting browser support it? its not tested in this poly either)
*    Providing demos from simple to complex with nested draggables/dropzones and the different combinations of effectAllowed and dropEffect.
*    Does anybody have an idea how we can introduce tests that aim to discover changes in browsers as quick as possible? SauceLabs is free for open source projects so if we are able to write some tests that could help in detecting breaking changes and keep us up-to-date about any browser changes would be quite nice.

JonDum commented:
@reppners 
> Detecting breaking changes is very important. Has anybody suggestions on how to tackle this? Write down a list of tests? Automated tests?

Wouldn't be too bad using nightwatch to connect to sauceLab's mobile selenium driver. The test would be something like:

```
'drag-drop' : function(browser) {
   //init stuff
   browser
   .moveToElement('.draggable', 0, 0)
   .mouseButtonDown(0)
   .moveToElement('.drag-zone', 10, 10)
   .mouseButtonUp(0)
   .asset(...)
}
```

Also, your `debug` stuff adds a lot of bloat to the script. Best practice is to use 
```
if(typeof DEBUG === 'undefined') {
   var DEBUG = false;
}
doSomethingNonDebug();
if(DEBUG) {
  //debug code here
}
```

so people with var replacement and dead-code removal can have that removed from the code upon minification. After running through a DefinePlugin and UglifyJS the above becomes:

```
if(typeof false === 'undefined') {
   var DEBUG = false;
}
doSomethingNonDebug();
if(false) {
   log('Debug thing')
}
```
Which then becomes

```
doSomethingNonDebug();
```


Minifiers won't touch `this.config.debug` because it has no way of knowing that's safe to remove. 


> Another thing I'm thinking about is modularity and how supporting browserify/webpack/etc. can be established as I have no experience on how this is done.

Polyfills generally don't need a UMD or shim since they don't need to export anything. `require()`ing the polyfill should run the polyfill and that's about it. If you want to get super fancy, I'd split the IsDragAndDropSupportedNatively into a separate module so you can require() that test and then require() the rest of the module in a separate async chunk/payload along with other polyfills.

Personally though, your code is super clean and well documented, but you went totally overboard with the TypeScript, build system, separate CSS files, embedded FontAwesome Icons, Utility classes

reppners commented:
@JonDum thx for your feedback. I realize that there is some bloat due to debug code and didn't know about the possibility to have dead code removed by UglifyJS and will try to get that incorporated. As a long term goal I would like to make the code simpler and more concise where possible without losing maintainability and of course functionality. Having tests is a precondition for starting work on this.

Sounds like setting up a demo page which can also be used as a test page is the way to go. Then define selenium tests to get the expected behavior covered.

I understand the added tools and separations can feel a bit "too much" but I really like maintainable, easy to contribute stuff and even more I like compilers yelling at me if I do something wrong. TypeScript does a pretty good job in giving support while still being able to leverage all the duck-typing freedom that JS offers.

timruffles commented:
@reppners I would also ask that we move as much as possible back to plain, free-standing function declarations. They're really easy to unit test and understand, unlike objects that need to be created in a certain way.

reppners commented:
Actually its 1,030 lines without comments when looking at the pure js. The ts-file is filled with comments and interfaces so its length is not really representable. I did keep all commented out parts of the spec included in case it would be ever needed to push support for spec-related things further. Also there is a lot of code related to "dropzone"-attribute support that maybe can be commented out until browsers start to support it if they will ever.

`DragAndDropInitializer`-class is purely static, listening on events that can start a drag operation and if an event turns out to not start a drag operation make the default behavior happen like click events or putting focus in an input field. There are some bugs left in this part, but I don't have much time left to spend on this atm. You are right about the config object not being needed in `TryFindDraggableTarget`, its a leftover that I did not remove yet. The `DragAndDropInitializer` also limits the simultaneous operations to one.

The `DragOperationController` is what was the DragDrop-class in your implementation. It implements the drag and drop processing model, fires events, contains instances of DataTransfer and related DragDataStore. It contains methods that implement various algorithms defined by the spec (`dragAndDropProcessModelIteration`, `DragOperationEnded`, `DetermineDropEffect`, `DetermineDragOperation`, `FindDropzoneElement`, `GetOperationForMatchingDropzone`). In addition there is code related to firing the events. That sums up to about 340 lines of pure js.

DataTransfer is implemented as class, as well as DragDataStore with the goal of being close to the spec. They take about 120 lines of pure js, most of it being logic in property getters/setters on the DataTransfer class.

About 130 lines of pure js-code are added for snapback transitions and scrolling when dragging close to edges. Thats code purely for "cosmetics" and usability.

So thats about 460 LOC spec related implementations, 130 LOC of new features regarding snapback and scrolling, 100 LOC on utils, and the rest is setup/cleanup/touch event handling/drag image handling.

Can you elaborate what you mean by plain, free-standing functions? I would like to keep a certain level of separation of concerns.

I'm totally open to get it more succinct but for the company project the primary goal was getting it working as defined by spec and desktop browser behavior so I did not put much effort in keeping it small in the first place.

reppners commented:
I took a stab at removing unneeded code, removing "dropzone" related stuff, make logging and debug code disappear on minification and refactor some things that we're not too involved in the dnd process model iteration. 

Its in a separate branch: https://github.com/reppners/ios-html5-drag-drop-shim/commit/0856424ed182ee7f06b3835040a4a72b97450eb9

before:
* js: 50.96KB (7.34KB gzipped), 1029 lines
* min.js: 25.02KB (5.74KB gzipped)

after:
* js: 37.79KB (6.01KB gzipped), 767 lines
* min.js: 16.88KB (4.07KB gzipped)

NoelAbrahams commented:
@reppners,

I'm going to try out your branch on some production code. Drag-and-drop is an essential feature of our app, hence this is very useful. (@timruffles's lib stopped working on iOS 8.4.1).

The size of the lib is a problem though.

The lib can be minified further by rewriting the private methods as simple functions. The problem is TypeScript puts them on the prototype, which means the minifier cannot tell if they are being used elsewhere. For example, there are 11 references to `dispatchDragEvent`.

Another alternative is to try the [Google Closure Compiler](https://closure-compiler.appspot.com/home) with advanced options - although some preparation will need to be done to ensure essential symbols and identifiers are preserved.

timruffles commented:
Yes - another reason I think it'd make a lot of sense to move all functions
(e.g not methods that need access to state) back be free standing again,
private to the module.

On Thu, Aug 27, 2015 at 11:17 AM, nabog <notifications@github.com> wrote:

> @reppners <https://github.com/reppners>,
>
> I'm going to try out your branch on some production code. Drag-and-drop is
> an essential feature of our app, hence this is very useful. (@timruffles
> <https://github.com/timruffles>'s lib stopped working on iOS 8.4.1).
>
> The size of the lib is a problem though.
>
> The lib can be minified further by rewriting the private methods as simple
> functions. The problem is TypeScript puts them on the prototype, which
> means the minifier cannot tell if they are being used elsewhere. For
> example, there are 11 references to dispatchDragEvent.
>
> Another alternative is to try the Google Closure Compiler
> <https://closure-compiler.appspot.com/home> with advanced options -
> although some preparation will need to be done to ensure essential symbols
> and identifiers are preserved.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/timruffles/ios-html5-drag-drop-shim/pull/44#issuecomment-135371894>
> .
>

reppners commented:
Makes sense. I will spend some time on this :)

@NoelAbrahams Further refactorings will land in the refactoring-branch mentioned in my previous comment. As far as iOS 8.4.1 support is concerned it should be working, but if not let me know so we can fix it.

timruffles commented:
As you've done a heap of work, I wanted to get involved a bit. I made a [quick stab](https://github.com/reppners/ios-html5-drag-drop-shim/compare/refactoring...timruffles:ts-privacy) at moving all methods that don't rely on internal state to stand-alone functions. Cuts down the compiled code size and simplifies the code, keeping stuff that potentially read/writes from mutable state to a minimum.

I've never written a line of TypeScript so this is probably a mess :)

One other request - if you're hoping to have the original maintainers involved - is to strip out the tautological @param style comments, e.g

```ts
    /** ...
        * @param e
        * @constructor
        */
    function OnTouchstart( e:TouchEvent ) {
```

It's noisy and kills readability for me. The JS community's style is very minimalist when it comes to commenting. Especially with TypeScript's richer type system, the Javadoc style comments don't seem justified.

I still feel there is a leaner, meaner version of the typescript code we could squeeze out. Right now I can't navigate the code to get a feel for it due to the comment blizzard :)

reppners commented:
Cool, thx :) 

> I've never written a line of TypeScript so this is probably a mess :)

No worries, TypeScript is only a superset of javascript so its all valid ;)

You're right about the comments containing too much noise, Those bits are auto-generated from Webstorm and I will strip them out. Also there is a lot of comments that contain "copy-paste" stuff from the html5 spec and I'd like to keep those things as a reference. Maybe they can be shortened to just reference/link to somewhere so the code is not hidden in a jungle of comments.

Apart from that there are those `fold`, `endfold` comment statements which act as collapsible markers and I hope every editor is capable of handling those. They act as a kind of a separator of different blocks to aid readability. Much like `#pragma mark` in objective-c or `region` in jetbrains-stuff like when doing Java or C#. They should help a lot to get an overview by first collapsing all and then step into those parts that are currently relevant to you.

reppners commented:
@timruffles I did another round of refactoring where I merged in your changes. Check the [refactoring branch](https://github.com/reppners/ios-html5-drag-drop-shim/tree/refactoring). Reported size of uglifyJS is 13836 bytes. Next up step would be to reduce the comment blizzard by removing spec related comments as much as possible by rather reference the spec instead of citing it. 

@NoelAbrahams You can try with the current state of the refactoring branch, the demo is fully working on my iPhone running iOS 8.4.1 - so it should also in a production app. DragImage not snapping back should not occur (anymore). Drag operations not finishing should absolutely not occur, but you need to provide more details on this if it is still happening.

NoelAbrahams commented:
@reppners,

The problems have not gone away with the latest in the refactoring branch.

[Here's a fiddle](https://jsfiddle.net/4aa2wr99/5/) that doesn't work in iOS 8.4.1. The drop operation never succeeds.

[Here's a fiddle](https://jsfiddle.net/exp42sq7/1/) using @timruffles lib that does work in iOS 8.4.1.

reppners commented:
@NoelAbrahams 
Thx for your feedback.
According to the spec you need to `preventDefault()` the `drop`-event. See the [updated fiddle](https://jsfiddle.net/steinharts/r1w96d28/)

Now the interesting part is that desktop browsers (tested on Safari and Chrome) seem to accept the `drop` without `preventDefault()` as soon as you access the dataTransfer prop. You can try it out by disabling the line where you access the dataTransfer-object in the `drop`-handler. Then the drag operation will silently fail/cancel.

Maybe the polyfill can be further be optimized to behave like desktop browsers, though thats not possible for every case. See a related issue https://github.com/marceljuenemann/angular-drag-and-drop-lists/pull/118

-------------------------------------------

I did further work on refactoring and minification and was able to reach around 10kB minified size. Will commit in the next few days since I just arrived from vacation.

NoelAbrahams commented:
@reppners,

That's helpful to know. However, in spite of setting `preventDefault()` on the `drop` event, the lib is not working as expected. I was trying to reproduce the problem, but I seem to have ended up with a red herring. I say this, because in our actual code `preventDefault` does exist on the `drop` event.

But let me try out your latest code and I'll get back with more details.

Also do you know if problem #48 has been resolved in your lib?

reppners commented:
@NoelAbrahams 
#48 should be resolved. Drag will only start after first `touch move` and if none happens the default action is initiated. However there still might be some bugs and cross-browser/platform issues on triggering the default action on input/form elements but simple clicks should be no problem.

NoelAbrahams commented:
@reppners,

I am experiencing the following problems (using the minified version dated 30-Aug-2015):

* The drag-and-drop stops working if the item is not dropped into the target. For example, if I have two items A and B, and A is touched and moved slightly, but not dropped into the drop target, then attempting to move A or B thereafter fails.
* After the failure described above, tapping on any element (e.g. a page tab) that causes the page to be redrawn (not reloaded, but redrawn in the SPA-sense, i.e. the DOM element containing A is removed dynamically), the image for A becomes detached and moves to the top left corner. Here is a screenshot of the effect:

![image](https://cloud.githubusercontent.com/assets/1106823/9836071/962bea56-5a02-11e5-81b0-451837035feb.png)

As far as I can tell, the drag-and-drop seems to be working correctly otherwise.

reppners commented:
@NoelAbrahams 
Must be a problem with the snapback transition not working correctly which then results in a hanging drag and drop operation, the operation never ends and the drag image keeps lingering around. Do you have the css file included containing the css rule for the snapback transition and can you tell if the rule is applied? I suspect the snapback transition that should be triggered by adding the css class containing the transition rule is never applied, therefore the `transitioned` event is never called and the drag operation keeps hanging around.

NoelAbrahams commented:
@reppners, I do have the CSS included. There is no conflict with the CSS classnames so it should be applied. I think this is more than just a CSS issue, because the drag stops working after the first time.

reppners commented:
@NoelAbrahams Can you try out the non-minified version? Because the minified version should be the same but it is rather untested and I was experimenting with minification config at that time.

reppners commented:
@NoelAbrahams So far I could not reproduce, minified version also seems not to be the culprit. Can you give me more info about the element you're dragging? Is it happening for every element in your app or just some specific kind of elements?

NoelAbrahams commented:
@reppners, the markup is similar to this

```html
<div class="det draggable" draggable="true">
  <div class="det-icon" >
    <figure class="oic draggable" draggable="true">
      <figcaption>
        <div >Institute Of Contemporary Arts</div>
     </figcaption>
     <img class="oic-image" draggable="false" src="http://example.com">
    </figure>
   </div>
</div>
```
And the CSS:

```css
.draggable {
    cursor: grab;
    cursor: -webkit-grab;
}

    .draggable:active {
        cursor: grabbing;
        cursor: -webkit-grabbing;
    }

.det {
    position: relative;
    width: 100%;
    outline: 0.1em solid #d9d9d9;
    background-color: #fff;
    box-sizing: border-box;
}

.det-icon {
    display: inline-block;
    width: 36%;
    padding: 1% 2%;
    box-sizing: border-box;
    height: 100%;
    min-height:210px;
}

.oic {
    width: 100%;
    height: 100%;
    cursor: pointer;
    font-size: 0.8em;
    text-align: center;
    position: relative;
}

.oic-image {
    width: 100%;
    padding: 5% 0 0 0;
}
```

The drag start and end events are bound like this

```TypeScript
root.addEventListener('dragstart', event => {

    this.isInProgress = true;  // Progress styling
    event.dataTransfer.effectAllowed = 'move';
    event.dataTransfer.setData('text', 'foo');
    
   this.onDragStart.trigger(event); // Custom event
});

root.addEventListener('dragend', event => {

    this.isInProgress = false; // Progress styling
    const dropOccurred = event.dataTransfer.dropEffect !== 'none';

    this.onDragEnd.trigger(event.create(dropOccurred)); // Custom event
});
```

Perhaps the point of interest here is that both '.det' and '.oic' are draggable. This works fine on desktop browsers.

reppners commented:
@NoelAbrahams
The issue you're experiencing I was not able to reproduce, though I had issues like you are describing whenever I had errors with the transition/transitionend. To debug this further I would ask you to include the non-minified script and look at the console output. This should at least give us a hint in what state the drag operation is hanging when it happens.

I put your markup into my development playground and pushed it with the recent changes to the gh-pages: http://reppners.github.io/ios-html5-drag-drop-shim/spec-compliance/

-----------------------------------------------------------------

The size now is at 10kB minified and around 3.4kB gzipped.

reppners commented:
@NoelAbrahams 
Finally was able to get hold of your issue. It happens for me when I activate the debug mode by defining DEBUG = true. This starts some transitions that are triggered in a fast sequence and cause the transitionEnd on the snapback transition to fail on the second drag operation. I hacked it by using setTimeout() instead of listening for the transitionEnd event and this way it works reliably. This lead me to the thought that you might do some kind of DOM manipulation in your drag and drop operations that also caused the transitionEnd to never get fired.

Also caught another bug when issuing the default action on an element when not doing a drag operation that caused the whole polyfill to stop working after it happened once.

https://github.com/reppners/ios-html5-drag-drop-shim/tree/refactoring

NoelAbrahams commented:
@reppners, that sounds rather promising.

I'll be able to try out the changes and provide some feedback by the end of this week. :+1:

NoelAbrahams commented:
@reppners,

We've been trying out the latest changes. I'm no longer seeing the problem, so that appears to be fixed. Here are some side-effects and related issues.

* The single tap no longer works. This should trigger the `click` event.
* The pinch zoom fails on draggable items. The item gets dragged by the action, rather than zooming.
* Sometimes I want to scroll the page by touching and moving the page down (or up). When the touched item is a draggable item this action is ineffective; the item gets dragged rather than the page scrolling.

I'm not sure if some of these issues can actually be resolved. For example for the last item, it's not clear how it will be possible to distinguish a scrolling action vs a dragging action.

These issues certainly affect the usability of our app, so if they cannot be fixed, we may have to reconsider the whole design.

reppners commented:
@NoelAbrahams 
The first two are bugs and we should be able to fix them. 

Is it the same elements that emit no `click` as you have posted above? On what exact element are you expecting the `click`? I never had any problems yet with the `click` event yet so I'm curious what the problem could be there.

For the pinch zoom I can filter the initial touch event on the number of touches so that should be easily fixed.

As for the last point, I think it is the application that has to provide some sort of drag handle so the user clearly knows where he can drag an item. It is certainly a problem, but none that I think should or can be fixed at the polyfill layer. Internet Explorer on MS Surface for example does Drag and Drop after a sustaining touch and you have to pull the element in a certain direction to make it work. I don't think the polyfill should do any of those things. The most it can do is providing an interface so an application can decide itself what kind of touch/gesture initiates a drag and drop operation. Didn't you have the same effect with the shim yet?

NoelAbrahams commented:
@reppners,

> Is it the same elements that emit no click as you have posted above

Yes. Also perhaps the issue may have been misunderstood: it's not the click event, it's that the `touch` event should trigger the `click` event. Obviously, that snippet is part of a bigger picture and it's possible that the interaction with other elements is causing the click to fail.

> I think it is the application that has to provide some sort of drag handle so the user clearly knows where he can drag an item. 

That's certainly a good suggestion and I think we'll look into that. One case that we have is of drag-and-drop being effected where the source is an image. When we have a page full of images it becomes impossible to scroll down. For this case if we can prevent drag being effected when pulled up or down (i.e. only works left or right) then that would solve the problem. I think we might look into doing that.

reppners commented:
@NoelAbrahams 
I pushed a commit to the refactoring branch that filters the initial touches to allow pinching and other gestures and hopefully more accurately detects when to initiate a default click.

I was not able to reproduce the missing `click` event. Can you provide a fiddle that demonstrates the bug?

NoelAbrahams commented:
@reppners, just to provide some feedback: the `click` even does not work still. I'm currently not able to spend time on reproducing the problem, because of other priorities. I hope to revisit this in December.

timruffles commented:
@xStrom  @jmeas what do you two think about this? I'm leaning towards merging in these changes, but using a cleaned up version of the compiled JS. I think it over-complicates such a simple project, and cuts down our contributor base too much.

xStrom commented:
I like the concept of making this library behave more closely to the spec. Most of my own work here has been towards that goal.

Regarding the specific code and the state that it is in, I haven't done a proper review/test yet. I should do that. I'm not a fan of regression bugs and wouldn't recommend merging unless we're reasonably sure there aren't any major regressions. I'll see if I can find the the time for a proper review this week.

Regarding TypeScript, I don't have any problems with it personally. I've used it in my own projects as well. Being a superset, it's much closer to JavaScript than something like CoffeeScript, so I think the effect on potential contributors shouldn't be too large. Of course some people may get confused, but the same could be said about prototype usage, as most JavaScript programmers are familiar with only some subset of the language. Editing functions with limited JavaScript knowledge would still be possible. I don't really have a strong opinion on this topic, but I like static typing, so if anything I'll lean a bit towards supporting TypeScript.

kamelkev commented:
Spent the last couple weeks evaluating various drag and drop libraries, including this one. I'm curious as to the planned direction regarding this pull request.

Specifically I noticed that this pull request seemingly takes the project from being implemented in javascript to being implemented in typescript. This is *not* a concern that I have, but I want to properly account for such complexity before making my own final selection.

I spent a good day reviewing the proposed branch from @reppners. I didn't notice any obvious regressions while using that branch vs the main branch, however I did notice that it was still not possible to drag items from a main document into an iFrame, something that I had hoped would magically start working once the library moved closer to the specification. Any comments there @reppners?

reppners commented:
@kamelkev 
I never tried to make it work with an iFrame. After taking a look at examples it seems there is nothing special to do when using the html5 DnD api so you are right in your assumption that it should just work. Unfortunately I don't have enough spare time to further investigate this right now. I'll try to spin up an example and take a look at it in the next few weeks.

Regarding typescript I think it doesn't add complexity. Its just ES6/ES7 features + type checking. In the long run its more self-documenting and easier to maintain. Comparing ES6 and TypeScript I'd say they are very close, its just adding (optional, flexible) type declarations.

kamelkev commented:
@reppners thanks for the follow up. Yeah I understand the nature of typescript, but I still need to account for the time in becoming familiar with it. I'm not against it if it improves the quality of a product, so non-issue there.

@timruffles Which direction are you planning? I see a note a few months ago, and seemingly nothing after that.

xStrom commented:
I think the most likely scenario is that some adaptation of this branch will go into master.

Unfortunately I still haven't managed to properly review it all. All the time I've had for this project has gone into maintaining the current master branch.

kamelkev commented:
Ok, that sounds promising. Thanks for the follow up.

timruffles commented:
Ok - does anyone have the time to push this forward? We'd need to roll in all the good stuff people have committed to the JS version too.

@reppners @xStrom

xStrom commented:
Time is certaintly scarce currently. However as you mentioned, the JS branch has been getting some nice polish over these last months and so the amount of work that needs to be done here keeps increasing. Thus it certainly makes sense to get this over with sooner rather than later.

I guess I can commit to doing a serious code review & parity check before the end of the month.

What's the decision on the TypeScript situation? Like I mentioned earlier, I don't have strong feelings here however I would rather keep it in TypeScript for the static typing benefits.

timruffles commented:
I'd be happy to go with TypeScript, especially now @reppners did such a great job on making a lean version.

@reppners any ideas on how to proceed?

timruffles commented:
BTW @reppners you're now a contributor - thanks again for doing such an awesome job with the TS code! :)

reppners commented:
> BTW @reppners you're now got collaborator/commit access - thanks again for doing such an awesome job with the TS code! :)

Thx! I really appreciate that you made me a collaborator for this project.

Much work did go in the [refactoring-branch](https://github.com/reppners/ios-html5-drag-drop-shim/tree/refactoring) so I think it makes sense to make a new pull request based on that branch.

To get all the stuff thats in the JS version I think we have to make a checklist to be sure we don't miss anything. I scanned the latest commits and merges and there were a few bugs fixed that we have to make sure are either not existing in the rewrite or will be fixed there as well.

So here is a checklist for things that I would like to do:

- [ ] use tsconfig.json for TS config instead of grunt task config
- [ ] introduce TSLint
- [ ] implement setDragImage() in rewrite
- [ ] make sure latest bugfixes either are not needed in first place or are applied to rewrite
- [ ] find solution for Safari 9 zooming the whole webpage when a drag item reaches right edge of the viewport (they changed something from v8 to v9 that I couldn't come up with a solution yet)
- [ ] refactor code to allow hook for customizing drag start conditions
- [ ] refactor code to allow hook for default action when drag was not started
- [ ] refactor code to allow hook for scrolling a container or viewport when dragging close to an edge

The last points became increasingly important to me. Instead of having the polyfill implement all kinds of use-cases that users of the polyfill might have a need for we should provide a few hooks so they can implement their own solutions. Even browsers are not behaving all the same when deciding on when to start a drag operation. Internet Explorer for example requires a sustaining touch and horizontal movement to start a drag operation.

If we are able to provide these hooks we can shrink the size of the library and everybody can implement it in a way that feels and works good for their specific product/use-case. Of course a certain level of basic functionality should be provided by the polyfill.

To get started we could merge the refactoring-branch to a new branch in this repository and work on from there until we are satisfied to make a beta-phase, where we try to fix all regression bugs that might occur.

Thoughts?

xStrom commented:
Merging *refactoring* as a new unstable branch here seems like a very good idea to me. We can then open separate issues for that branch and save on some potential confusion.

Create a new pull request that does this merger as a new branch, and lets get this party started.

reppners commented:
closing this in favor of #61
